// @ark-ui/react/dist/components/marquee/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const marqueeContent = require('./marquee-content.cjs');
const marqueeContext = require('./marquee-context.cjs');
const marqueeEdge = require('./marquee-edge.cjs');
const marqueeItem = require('./marquee-item.cjs');
const marqueeRoot = require('./marquee-root.cjs');
const marqueeRootProvider = require('./marquee-root-provider.cjs');
const marqueeViewport = require('./marquee-viewport.cjs');
const useMarquee = require('./use-marquee.cjs');
const useMarqueeContext = require('./use-marquee-context.cjs');
const marquee$1 = require('./marquee.cjs');
const marquee = require('@zag-js/marquee');



exports.MarqueeContent = marqueeContent.MarqueeContent;
exports.MarqueeContext = marqueeContext.MarqueeContext;
exports.MarqueeEdge = marqueeEdge.MarqueeEdge;
exports.MarqueeItem = marqueeItem.MarqueeItem;
exports.MarqueeRoot = marqueeRoot.MarqueeRoot;
exports.MarqueeRootProvider = marqueeRootProvider.MarqueeRootProvider;
exports.MarqueeViewport = marqueeViewport.MarqueeViewport;
exports.useMarquee = useMarquee.useMarquee;
exports.useMarqueeContext = useMarqueeContext.useMarqueeContext;
exports.Marquee = marquee$1;
Object.defineProperty(exports, "marqueeAnatomy", {
  enumerable: true,
  get: () => marquee.anatomy
});
