// @ark-ui/react/dist/components/splitter/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const splitter = require('@zag-js/splitter');
const splitterContext = require('./splitter-context.cjs');
const splitterPanel = require('./splitter-panel.cjs');
const splitterResizeTrigger = require('./splitter-resize-trigger.cjs');
const splitterResizeTriggerIndicator = require('./splitter-resize-trigger-indicator.cjs');
const splitterRoot = require('./splitter-root.cjs');
const splitterRootProvider = require('./splitter-root-provider.cjs');
const useSplitter = require('./use-splitter.cjs');
const useSplitterContext = require('./use-splitter-context.cjs');
const splitter$1 = require('./splitter.cjs');



Object.defineProperty(exports, "getSplitterLayout", {
  enumerable: true,
  get: () => splitter.layout
});
Object.defineProperty(exports, "splitterAnatomy", {
  enumerable: true,
  get: () => splitter.anatomy
});
exports.SplitterContext = splitterContext.SplitterContext;
exports.SplitterPanel = splitterPanel.SplitterPanel;
exports.SplitterResizeTrigger = splitterResizeTrigger.SplitterResizeTrigger;
exports.SplitterResizeTriggerIndicator = splitterResizeTriggerIndicator.SplitterResizeTriggerIndicator;
exports.SplitterRoot = splitterRoot.SplitterRoot;
exports.SplitterRootProvider = splitterRootProvider.SplitterRootProvider;
exports.useSplitter = useSplitter.useSplitter;
exports.useSplitterContext = useSplitterContext.useSplitterContext;
exports.Splitter = splitter$1;
