// @ark-ui/react/dist/components/splitter/splitter.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const splitter = require('@zag-js/splitter');
const splitterContext = require('./splitter-context.cjs');
const splitterPanel = require('./splitter-panel.cjs');
const splitterResizeTrigger = require('./splitter-resize-trigger.cjs');
const splitterResizeTriggerIndicator = require('./splitter-resize-trigger-indicator.cjs');
const splitterRoot = require('./splitter-root.cjs');
const splitterRootProvider = require('./splitter-root-provider.cjs');



Object.defineProperty(exports, "getLayout", {
  enumerable: true,
  get: () => splitter.layout
});
exports.Context = splitterContext.SplitterContext;
exports.Panel = splitterPanel.SplitterPanel;
exports.ResizeTrigger = splitterResizeTrigger.SplitterResizeTrigger;
exports.ResizeTriggerIndicator = splitterResizeTriggerIndicator.SplitterResizeTriggerIndicator;
exports.Root = splitterRoot.SplitterRoot;
exports.RootProvider = splitterRootProvider.SplitterRootProvider;
